mesos-client/src/main/scala/com/mesosphere/mesos/client/MasterDetector.scala

```diff
@@ -0,0 +1,126 @@
+package com.mesosphere.mesos.client
+import java.net.URL
+import java.util.concurrent.CompletionStage
+
+import com.typesafe.scalalogging.StrictLogging
+import org.apache.curator.framework.CuratorFrameworkFactory
+import org.apache.curator.retry.RetryOneTime
+import org.apache.curator.x.async.AsyncCuratorFramework
+import org.apache.mesos.v1.Protos
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+
+import scala.async.Async.{async, await}
+import scala.collection.JavaConverters._
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+trait MasterDetector {
+
+  /** @return the Mesos master URL for the cluster. */
+  def getMaster()(implicit ex: ExecutionContext): CompletionStage[URL]
+
+  /** @return whether the master string is valid. */
+  def isValid(): Boolean
+}
+
+object MasterDetector {
+
+  /**
+    * Constructs a master detector base on the master string passed.
+    *
+    * @param master The master string should be one of:
+    *               host:port
+    *               http://host:port
+    *               zk://host1:port1,host2:port2,.../path
+    *               zk://username:password@host1:port1,host2:port2,.../path
+    * @return A master detector.
+    */
+  def apply(master: String): MasterDetector = {
+    if (master.startsWith("zk://")) {
+      Zookeeper(master)
+    } else {
+      Standalone(master)
+    }
+  }
+}
+
+case class Zookeeper(master: String) extends MasterDetector with StrictLogging {
+  require(master.startsWith("zk://"), s"$master does not start with zk://")
+
+  case class ZkUrl(auth: Option[String], servers: String, path: String)
+
+  implicit val mesosInfoRead: Reads[Protos.MasterInfo] = (
+    (JsPath \ "hostname").read[String] ~
+      (JsPath \ "port").read[Int] ~
+      (JsPath \ "id").read[String] ~
+      (JsPath \ "ip").read[Int]
+  ) { (hostname, port, id, ip) =>
+    Protos.MasterInfo.newBuilder().setHostname(hostname).setPort(port).setId(id).setIp(ip).build()
+  }
+
+  override def isValid(): Boolean = Try(parse()).map(_ => true).getOrElse(false)
+
+  override def getMaster()(implicit ex: ExecutionContext): CompletionStage[URL] = {
```

Comment:
Why did you reimplement all this ZK connection/reading logic here? All of that already exists in `persistence-zookeeper` module and can look like:
```
  val client: CuratorFramework = zkClient()
  val factory: AsyncCuratorBuilderFactory = AsyncCuratorBuilderFactory(client)
  val store: ZooKeeperPersistenceStore = new ZooKeeperPersistenceStore(DummyMetrics, factory, parallelism = 1)

  val bytes = await(store.read(Node(leaderPath)))
```

Reply:
I'm not sure I want to pull in `usi-persistence` and `usi-persistence-zookeeper`. I was actually hoping that there is a Curator recipe but I only found  [LeaderElection](https://curator.apache.org/curator-recipes/leader-election.html) which seems to participate in the election and is not a read only. Maybe @timcharper knows something.

Reply:
> 'm not sure I want to pull in `usi-persistence` and `usi-persistence-zookeeper`.

I agree. That's why I would argue that this `MasterDetector` belongs someplace else.

Reply:
Additionally: though thread-safe, it is considered _a bad practice to have > 1 zkClients_ in the project. It should be passed to the `MasterDetector`.

Reply:
> you're basically just reading a node.

The children sorted and then one node. I wonder if this should in in an transaction  so that there is no new leader.

Reply:
You can use `ZookeeperPersistenceStore` to read multiple nodes (and/or children) in a transaction. But you **can't** have this `filter.min` logic in one:
```val children = await(asyncClient.getChildren.forPath(path).toScala).asScala
      val leader = children.filter(_.startsWith("json.info")).min```

Reply:
I moved it out. This is still ugly. We'd have to inject the client which I don't really like. Maybe I should just drop this and move the code into the Mesos plugin.